crates/storage/trie/db.rs

```diff
@@ -0,0 +1,137 @@
+use crate::error::StoreError;
+use ethereum_rust_core::rlp::{decode::RLPDecode, encode::RLPEncode};
+use ethereum_types::H256;
+use libmdbx::{
+    orm::{table, Database},
+    table_info,
+};
+
+/// Libmbdx database representing trie state
+/// It contains two tables, one mapping node references to rlp encoded nodes, and another one mapping node hashes to node references
+/// All nodes are stored in the DB and no node is ever removed
+/// Only node hahses for root nodes computed with `Trie::compute_hash` are stored
+/// Contains the next node reference to be used when storing a new node
+use super::{node::Node, node_ref::NodeRef};
+pub struct TrieDB {
+    db: Database,
+    next_node_ref: NodeRef,
+}
+
+/// RLP-encoded trie node
+pub type NodeRLP = Vec<u8>;
+/// RLP-encoded node hash
+pub type NodeHashRLP = [u8; 32];
+
+table!(
+    /// NodeRef to Node table
+    ( Nodes ) NodeRef => NodeRLP
+);
+
+table!(
+    /// NodeHash to NodeRef table
+    /// Stores root nodes which's hashes have been computed by `compute_hash`
+    ( RootNodes ) NodeHashRLP => NodeRef
+);
+
+impl TrieDB {
+    /// Opens a DB created by a previous execution or creates a new one if it doesn't exist
+    pub fn init(trie_dir: &str) -> Result<(TrieDB, Option<NodeRef>), StoreError> {
+        TrieDB::open(trie_dir).or_else(|_| TrieDB::create(trie_dir).map(|db| (db, None)))
+    }
+
+    /// Creates a new clean DB
+    pub fn create(trie_dir: &str) -> Result<TrieDB, StoreError> {
+        let tables = [table_info!(Nodes), table_info!(RootNodes)]
+            .into_iter()
+            .collect();
+        let path = Some(trie_dir.into());
+        Ok(TrieDB {
+            db: Database::create(path, &tables).map_err(StoreError::LibmdbxError)?,
+            next_node_ref: NodeRef::new(0),
+        })
+    }
+
+    /// Opens a DB created by a previous execution
+    /// Also returns root node reference if available
+    pub fn open(trie_dir: &str) -> Result<(TrieDB, Option<NodeRef>), StoreError> {
+        // Open DB
+        let tables = [table_info!(Nodes), table_info!(RootNodes)]
+            .into_iter()
+            .collect();
+        let db = Database::open(trie_dir, &tables).map_err(StoreError::LibmdbxError)?;
+        // Set the next node reference based on the last stored node
+        let last_node_ref = last_node_ref(&db)?;
+        let next_node_ref = last_node_ref.map(|nr| nr.next()).unwrap_or_default();
+        Ok((TrieDB { db, next_node_ref }, last_node_ref))
+    }
+
+    /// Retrieves a node based on its reference
+    pub fn get_node(&self, node_ref: NodeRef) -> Result<Option<Node>, StoreError> {
```

Comment:
shouldn't `NodeRef` be a implementation details and not exposed in a public function?

Reply:
The db module is not public, NodeRef is an implementation detail of the trie module only public within the trie module